src/runtime-rs/crates/runtimes/common/src/runtimes_utils/host_hooks.rs

```diff
@@ -0,0 +1,371 @@
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let state = serde_json::to_string(st)?;
+    let join_handle = tokio::spawn(async move {
+        // write state to stdin of child process
+        if let Some(mut stdin) = child.stdin.take() {
+            match stdin.write_all(state.as_bytes()).await {
+                Ok(_) => {}
+                Err(e) => {
+                    info!(sl!(), "write to child stdin error: {:?}", e);
+                }
+            }
+        }
+
+        // read info from stdout of child process
+        if let Some(stdout) = child.stdout.as_mut() {
+            let mut out = String::new();
+            match stdout.read_to_string(&mut out).await {
+                Ok(_) => {
+                    info!(sl!(), "child stdout: {:?}", out.as_str());
+                }
+                Err(e) => {
+                    info!(sl!(), "read from child stdout failed: {:?}", e);
+                }
+            }
+        }
+
+        // read err info from stderr of child process
+        let mut err = String::new();
+        if let Some(stderr) = child.stderr.as_mut() {
+            match stderr.read_to_string(&mut err).await {
+                Ok(_) => {
+                    info!(sl!(), "child stderr: {:?}", err.as_str());
+                }
+                Err(e) => {
+                    info!(sl!(), "read from child stderr failed: {:?}", e);
+                }
+            }
+        }
+
+        match child.wait().await {
+            Ok(exit_status) => {
+                let exit_code = exit_status
+                    .code()
+                    .ok_or_else(|| anyhow!("hook exit status has no code"))?;
+                if exit_code != 0 {
+                    error!(
+                        sl!(),
+                        "hook {} exit code: {}, error message: {}",
+                        &binary_path.display(),
+                        exit_code,
+                        err
+                    );
+                    return Err(anyhow!(nix::Error::UnknownErrno));
+                }
+
+                debug!(sl!(), "hook {} exit code: 0", &binary_path.display());
+                Ok(())
+            }
+            Err(e) => Err(anyhow!(
+                "wait child process error: {} {}",
+                e,
+                e.raw_os_error().unwrap()
+            )),
+        }
+    });
+
+    match tokio::time::timeout(execution_timeout, join_handle).await {
+        Ok(r) => r.unwrap(),
+        Err(_) => Err(anyhow!(nix::Error::ETIMEDOUT)),
+    }
+}
+
+// check and reserve valid environment variables
+// invalid env var may cause panic, refer to https://doc.rust-lang.org/std/env/fn.set_var.html#panics
+// key should not:
+//  * contain NUL character '\0'
+//  * contain ASCII equal sign '='
+//  * be empty
+// value should not:
+//  * contain NUL character '\0'
+fn valid_env(e: &str) -> Option<(&str, &str)> {
```

Comment:
Agent has the same code, and should use a common function laid in `src/libs`.

But you can leave it in another PR.

Reply:
I have found that there is also some code implementing `execute_hooks` in `src/libs`:
https://github.com/kata-containers/kata-containers/blob/main/src/libs/kata-sys-util/src/hooks.rs

Maybe I will use that code to refactor this PR. Or before that, also refactor the agent's hook execution with the same code (: